libs/wingc/src/lsp/completions.rs

```diff
@@ -102,29 +24,428 @@ pub unsafe extern "C" fn wingc_on_completion(ptr: u32, len: u32) -> u64 {
+		let found_env = found_scope.env.borrow();
+		let found_env = found_env.as_ref().expect("Scope should have an env");
+		let found_stmt_index = scope_visitor
+			.found_stmt_index
+			.expect("Found scope should have a statement index");
+
+		let mut completions = vec![];
+
+		for symbol_data in found_env.symbol_map.iter().filter(|s| {
+			if let StatementIdx::Index(i) = s.1 .0 {
+				// within the found scope, we only want to show symbols that were defined before the current position
+				i <= found_stmt_index
+			} else {
+				true
+			}
+		}) {
+			let symbol_kind = &symbol_data.1 .1;
+
+			completions.push(format_symbol_kind_as_completion(symbol_data.0, symbol_kind));
+		}
+
+		// The following iteration logic is needed due to a bug:
+		// The root environment is not properly accessible from .parent
+		// https://github.com/winglang/wing/issues/1644
+		// So we can't use the normal iteration logic with ancestry
+		let mut parent_env = found_env.parent;
+		while let Some(current_env) = parent_env {
+			if current_env.is_root() {
+				for symbol_data in root_env.symbol_map.iter() {
+					completions.push(format_symbol_kind_as_completion(symbol_data.0, &symbol_data.1 .1));
+				}
+				break;
+			} else {
+				for data in current_env.iter(false) {
+					completions.push(format_symbol_kind_as_completion(&data.0, &data.1));
+				}
+			}
+			parent_env = current_env.parent;
+		}
+
+		completions
+	}))
+}
+
+/// Gets accessible properties on a type as a list of CompletionItems
+fn get_completions_from_type(
+	type_: &UnsafeRef<Type>,
+	types: &Types,
+	current_phase: Option<Phase>,
+	is_instance: bool,
+) -> Vec<CompletionItem> {
+	match &**type_ {
+		Type::Class(c) => get_completions_from_class(c, current_phase, is_instance),
+		Type::Resource(c) => get_completions_from_class(c, current_phase, is_instance),
+		Type::Interface(i) => get_completions_from_class(i, current_phase, is_instance),
+		Type::Struct(s) => get_completions_from_class(s, current_phase, is_instance),
+		Type::Enum(enum_) => {
+			let variants = &enum_.values;
+			variants
+				.iter()
+				.map(|item| CompletionItem {
+					label: item.name.clone(),
+					detail: Some(enum_.name.name.clone()),
+					kind: Some(CompletionItemKind::ENUM_MEMBER),
+					..Default::default()
 				})
-				.collect(),
-		)
-	})
+				.collect()
+		}
+		Type::Optional(t) => get_completions_from_type(t, types, current_phase, is_instance),
+		Type::Void | Type::Function(_) | Type::Anything => vec![],
+		Type::Number
+		| Type::String
+		| Type::Duration
+		| Type::Boolean
+		| Type::Json
+		| Type::MutJson
+		| Type::Array(_)
+		| Type::MutArray(_)
+		| Type::Map(_)
+		| Type::MutMap(_)
+		| Type::Set(_)
+		| Type::MutSet(_) => {
+			// This section of the code is hacky
+			// This is needed because our builtin types have no API
+			// So we have to get the API from the std lib
+			// But the std lib sometimes doesn't have the same names as the builtin types
```

Comment:
What changes would be needed from the compiler in order to avoid this hacky code? Can we create or link to an issue?

Reply:
Honestly, I'm not sure exactly what needs to happen. Here's what I could think of https://github.com/winglang/wing/issues/1780

I'll add it to the comment